spinalcordtoolbox/deepseg/inference.py

```diff
@@ -77,3 +91,158 @@ def segment_and_average_volumes(model_paths, input_filenames, options):
+    start = time.time()
+
+    # run inference
+    test_input = batch["image"].to(torch.device("cpu"))
+    batch["pred"] = sliding_window_inference(test_input, inference_roi_size, mode="gaussian",
+                                             sw_batch_size=4, predictor=predictor, overlap=0.5, progress=False)
+    pred = ds_monai.postprocessing(batch, test_post_pred)
+
+    end = time.time()
+    print('Inference done.')
+    total_time = end - start
+    print(f'Total inference time: {int(total_time // 60)} minute(s) {int(round(total_time % 60))} seconds')
+
+    # this takes about 0.25s on average on a CPU
+    # image saver class
+    _, fname, ext = extract_fname(path_img)
+    postfix = "seg"
+    target = f"_{postfix}"
+    pred_saver = SaveImage(
+        output_dir=tmpdir, output_postfix=postfix, output_ext=ext,
+        separate_folder=False, print_log=False)
+    # save the prediction
+    fname_out = os.path.join(tmpdir, f"{fname}_{postfix}{ext}")
+    logger.info(f"Saving results to: {fname_out}")
+    pred_saver(pred)
+
+    return [fname_out], [target]
+
+
+def segment_nnunet(path_img, tmpdir, predictor):
+    """
+    This script is used to run inference on a single subject using a nnUNetV2 model.
+
+    Author: Jan Valosek, Naga Karthik
+    Original script: https://github.com/ivadomed/model_seg_sci/blob/4184bc22ef7317b3de5f85dee28449d6f381c984/packaging/run_inference_single_subject.py
+    """
+    # Copy the file to the temporary directory using shutil.copyfile
+    path_img_tmp = os.path.join(tmpdir, os.path.basename(path_img))
+    shutil.copyfile(path_img, path_img_tmp)
+    logger.info(f'Copied {path_img} to {path_img_tmp}')
+
+    # Get the original orientation of the image, for example LPI
+    orig_orientation = get_orientation(Image(path_img_tmp))
+
+    # Get the orientation used by the model
+    if "SCI" in predictor.plans_manager.dataset_name:
+        model_orientation = "RPI"
+    else:
+        assert "levels" in predictor.plans_manager.dataset_name  # rootlets dataset
+        model_orientation = "LPI"
+
+    # Reorient the image to model orientation if not already
+    img_in = Image(path_img_tmp)
+    if orig_orientation != model_orientation:
+        logger.info(f'Changing orientation of the input to the model orientation ({model_orientation})...')
+        img_in.change_orientation(model_orientation)
+
+    # Create directory for nnUNet prediction
+    tmpdir_nnunet = os.path.join(tmpdir, 'nnUNet_prediction')
+    fname_prediction = os.path.join(tmpdir_nnunet, os.path.basename(add_suffix(path_img_tmp, "_pred")))
+    os.mkdir(tmpdir_nnunet)
+
+    # Run nnUNet prediction
+    print('Starting inference...')
+    start = time.time()
+
+    data = np.expand_dims(img_in.data, axis=0).transpose([0, 3, 2, 1]).astype(np.float32)
+    pred = predictor.predict_single_npy_array(
+        input_image=data,
+        image_properties={'spacing': img_in.dim[4:7]},
+    ).transpose([2, 1, 0])
```

Comment:
In order to fix a `Bus Error` on our CentOS Stream 8 runner, I needed to replace a call to [`predictor.predict_from_files`](https://github.com/ivadomed/model_seg_sci/blob/4184bc22ef7317b3de5f85dee28449d6f381c984/packaging/run_inference_single_subject.py#L186) with `predictor.predict_single_npy_array`. (My hunch is that the Bus Error was due to the fact that the former function uses `multiprocessing`.)

However, when I did so, I found that numpy array was misoriented, and I had to perform these `transpose` commands to replicate the old behavior, which just feels... wrong?

I'm wondering whether there's a way to preprocess the data such that these transpose commands aren't necessary. So, I would love some input from reviewers on this. :)

Reply:
((Also, I'm curious to benchmark the difference between the two `predict` functions to see if the update has any effect.))

Reply:
I agree with your feeling that having some extra `transpose`s seems fishy. I tried digging into the nnUnet code to see where the difference between `predict_from_files` and `predict_sungle_npy_array` is, but so far I haven't found the actual bit of code that does anything with re-orientation or array indices. Maybe I should look into [this bit of documentation](https://github.com/MIC-DKFZ/nnUNet/blob/master/documentation/dataset_format.md#supported-file-formats) and [that bit of documentation](https://github.com/MIC-DKFZ/nnUNet/blob/master/documentation/dataset_format_inference.md) next.

I also agree with your guess that `multiprocessing` is a potential source of problems, so it's good to avoid it here.

Another reason to try to understand why we're `transpose`ing: do we also need to re-order the `img_in.dim[4:7]`?

Reply:
Just a little throwaway comment: It may have something to do with different origins for NumPy arrays vs. Nibabel/SimpleITK images? (I feel like I've run into that before -- NumPy being bottom-left and Nibabel being top-left?) 

Actually wait, no, that wouldn't explain the reorientation of the axes, but... "framework differences" may be a possible angle to consider if we can't find any explicit reorientation differences in the upstream functions.

I can come back to this to explore it more deeply later, but I wanted to make a quick comment while it was fresh in my mind.

Reply:
Testing both `predict_from_files` and `predict_single_npy_array` using the rootlets test:

<details>
<summary> Long investigation notes </summary>

- The input image, `t2.nii.gz`, is originally in AIL orientation (dim from `sct_image -header`: `[60, 55, 52]`):

![image](https://github.com/spinalcordtoolbox/spinalcordtoolbox/assets/16181459/e24ec6dd-d574-4813-9883-4f4f1f635a4d)

- However, it is reoriented to LPI prior to predictions (dim: `[52, 60, 55]`):

![image](https://github.com/spinalcordtoolbox/spinalcordtoolbox/assets/16181459/64014c5e-f6d1-4dfd-a133-bced091e1a19)

- In the `predict_from_files` case:
    - The filename of the reoriented image is passed to the function. 
    - (It is unclear to me how nnUNet loads the image from the filename.)
- In the `predict_single_npy_array` case:
    - The reoriented image is loaded into an `Image` object, and the numpy array is accessed via `Image.data`, where it has the dimensions `[52, 60, 55]` (same output as `sct_image -header`).
    - The transpose command changes the image from `[52, 60, 55]` to `[55, 60, 52]`. (This is equivalent to an IPL image, which is a very strange orientation that is seemingly independent of both the original and reoriented orientations!)
    - To be clear: If I remove the transpose commands, and pass in the LPI image array as-is, then the output numpy array will be empty.

So, I can't spot any issues that might have come from SCT's code (e.g. dependence on input orientation, accidentally passing the wrong image, etc.)

Digging into nnUNet's source code:

- I notice that the `predict_single_npy_array` applies preprocessing via the following trace:
    - `ppa = PreprocessAdapterFromNpy`
    - `preprocessor = configuration_manager.preprocessor_class` (assigned in __init__.py)
    - `next(ppa)` -> `generate_train_batch`
    - `preprocessor.run_case_npy`
- By comparison, `predict_from_files` applies preprocessing via the following trace: 
    - `self._internal_get_data_iterator_from_lists_of_filenames`
    - `preprocessing_iterator_fromfiles`
    - `preprocess_fromfiles_save_to_queue`
    - `preprocessor = configuration_manager.preprocessor_class`
    - `preprocessor.run_case`
    - `preprocessor.run_case_npy`
- So, both prediction functions arrive at the same preprocessing method. The shapes of the I/O for `run_case_npy` method is:
    - `predict_from_files`: `(55, 60, 52)` -> `(69, 75, 65)`
    - `predict_single_numpy_array` (with transpose): `(69, 75, 65)`
    - `predict_single_numpy_array` (no transpose): `(52, 60, 55)` -> `(65, 75, 69)`
- Aha! So, there's the smoking gun: When passing the reoriented LPI image filepath to `predict_from_files`, it gets loaded in that strange IPL orientation _before preprocessing even occurs_. (Thus, the transposing simply matches how nnUNet loads images from filepaths.)

The next question is: Why does `predict_from_files` load LPI images as IPL?

- Inside the stack trace, within the `run_case` function, images are loaded via `rw.read_images`.
- `read_images` is an abstract method that, for this case, manifests as `nnunetv2.imageio.simpleitk_reader_writer.SimpleITKIO`.
- This method calls `sitk.ReadImage(f)` internally. When looking at the output of this function, the metadata shows the LPI orientation (`[52, 60, 55]`)
- However, later on, `sitk.GetArrayFromImage(itk_image)` is called. And _this_ returns an array of size `[55, 60, 52]`.)
- This seems to be backed up by...

</details>

The [documentation for SimpleITK](http://insightsoftwareconsortium.github.io/SimpleITK-Notebooks/Python_html/03_Image_Details.html#Conversion-between-numpy-and-SimpleITK):

> SimpleITK and numpy indexing access is in opposite order!
>
> SimpleITK: image[x,y,z]
> numpy: image_numpy_array[z,y,x]

_What?_

tl;dr: nnUNet loads images from files using SimpleITK. When fetching a numpy array from a SimpleITK image, the axes are swapped. But, when fetching a numpy array from a nibabel image (used by SCT), the axes are _not_ swapped. So, the transpose commands are needed to match SimpleITK's axes-swapping behavior that nnUNet inherits.

Trying to learn why SimpleITK implemented things this way leads to discussions about [low-level efficiency and Fortan/C axes ordering](https://discourse.itk.org/t/why-to-invert-the-order-of-the-indices-between-numpy-and-simpleitk/809/15). tl;dr: The first axis is the "fast axis" in NumPy, so SimpleITK chooses to rehspae `[x,y,z]` images to `[z,y,x]` NumPy arrays for efficiency reasons. 

This isn't intuitive at all! But, at least I can document why the transposes are necessary.

> Side note: Does this mean that technically, all of our nnUNet models were trained... not in RPI/LPI, but in IPL/IPR???

Reply:
> Hmm, is there a way to verify this?

The closest thing I can think of for verification would be to check the code used for training to see if it uses `sitk.GetArrayFromImage(itk_image)`. I looked at the [`training/`](https://github.com/ivadomed/model-spinal-rootlets/tree/main/training) folder, but couldn't find any scripts that call `nnunetv2`. I might be missing something obvious, though?

Reply:
You're not missing anything. The nnunet training is done using a single bash command, `nnunetv2_train` (as documented, for example, [here](https://github.com/ivadomed/model-spinal-rootlets/pull/23#issuecomment-1888047165)). This is why there is no training script under the [training/](https://github.com/ivadomed/model-spinal-rootlets/tree/main/training) folder. Alternatively, the `nnunetv2_train` command can be wrapped into a simple wrapper script: [example](https://github.com/ivadomed/utilities/blob/main/training_scripts/run_nnunet.sh).

Reply:
So, it seems like how nnUNetv2 loads images during training is determined by the `overwrite_image_reader_writer` key in the [`dataset.json`](https://github.com/MIC-DKFZ/nnUNet/blob/master/documentation/dataset_format.md#datasetjson) configuration. This key is optional, and if not provided, nnUNet will determine the reader automatically for the user based on the image file extension of the data. 

Note that nnUNetv2 [_does_ support `nibabel`](https://github.com/MIC-DKFZ/nnUNet/blob/997804c7510634dc8fd83f1194b434c60815a93e/documentation/dataset_format.md?plain=1#L63-L65) as an image reader. However, I think the order of precedence is set by [this list](https://github.com/MIC-DKFZ/nnUNet/blob/997804c7510634dc8fd83f1194b434c60815a93e/nnunetv2/imageio/reader_writer_registry.py#L14-L20) -- since SimpleITKIO comes before NibabelIO, I think it gets used by default for `.nii.gz` images. 

In other words, to get the Nibabel axis behavior, we would have needed to explicitly specify `NibabelIO` in the `dataset.json` file. And, since we didn't, `SimpleITKIO` got used by default, resulting in the axis swap behavior during training.

Reply:
Reported upstream here: https://github.com/MIC-DKFZ/nnUNet/issues/1951

Reply:
Thanks for this amazing investigation, @joshuacwnewton ! All our scripts were depending on I/O from the nnunet side which always ensured that the input shapes/orientations during training and inference were the same. Hence, we were never aware that this was happening behind the scenes! 

To add my two cents to this investigation: 

> all of our nnUNet models were trained... not in RPI/LPI, but in IPL/IPR???

This very much seems like the case! nnUNet always does a "[fingerprint extraction](https://github.com/MIC-DKFZ/nnUNet/blob/master/documentation/how_to_use_nnunet.md#experiment-planning-and-preprocessing)" step where all the information regarding the dataset is gathered. [This line](https://github.com/MIC-DKFZ/nnUNet/blob/master/nnunetv2/experiment_planning/dataset_fingerprint/fingerprint_extractor.py#L115) determines which image reader/writer is used. Now, because our dataset-conversion scripts (SCIseg model [example](https://github.com/ivadomed/model_seg_sci/blob/main/dataset-conversion/convert_bids_to_nnUNetv2_all_sci_data.py#L562)) do not explicitly specify what IO reader to use, the reader is determined automatically by nnunet via the `file_ending` key in this [function](https://github.com/MIC-DKFZ/nnUNet/blob/master/nnunetv2/imageio/reader_writer_registry.py#L41). It is here that the [list](https://github.com/MIC-DKFZ/nnUNet/blob/master/nnunetv2/imageio/reader_writer_registry.py#L14-L19) of IO readers is followed as per that order (which you already found out above).

Reply:
Adding this just to document -- It seems that even if we use NibabelIO, the [transpose operation](https://github.com/MIC-DKFZ/nnUNet/blob/997804c7510634dc8fd83f1194b434c60815a93e/nnunetv2/imageio/nibabel_reader_writer.py#L54-L55) is still performed just to fit into the nnunet structure

Reply:
Ah! Thanks so much @naga-karthik ! A few lines above your link, I think [this code](https://github.com/MIC-DKFZ/nnUNet/blob/997804c7510634dc8fd83f1194b434c60815a93e/nnunetv2/imageio/nibabel_reader_writer.py#L49-L52) reverses the `spacing`, so that means we need to also reverse the axes in `image_properties`.